{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e103e923",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressWarnings(require('pkgmaker',quietly = T))\n",
    "require('plyr',quietly = T)\n",
    "l_ply(c('dplyr',\n",
    "        'plotly',\n",
    "        'data.table',\n",
    "#         'jsonlite', \n",
    "#         'rjson',\n",
    "#         'httr',\n",
    "        'tidyr',  \n",
    "        'reshape2',\n",
    "        'factoextra',\n",
    "        'kableExtra',\n",
    "        'parallel',\n",
    "        'doParallel',\n",
    "        'scales',\n",
    "        'StatMatch',\n",
    "        'lattice',\n",
    "        'utils',\n",
    "        'missMDA',\n",
    "        'RMariaDB',\n",
    "        'pacman',\n",
    "        'parmigene',\n",
    "        'matrixTests',\n",
    "        'hablar',\n",
    "        'tibble', \n",
    "        'DEqMS', \n",
    "        'pcaMethods'), function(pkg) {\n",
    "          invisible(capture.output(irequire(pkg, quiet = T, autoinstall = T)))\n",
    "        })"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4d52db7",
   "metadata": {},
   "source": [
    "## Example data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9d01bcff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# url <- \"https://ftp.ebi.ac.uk/pride-archive/2016/06/PXD004163/Yan_miR_Protein_table.flatprottable.txt\"\n",
    "# download.file(url, destfile = \"./data/miR_Proteintable.txt\",method= \"auto\")\n",
    "\n",
    "df.prot = read.table(\"./data/miR_Proteintable.txt\",stringsAsFactors = FALSE,\n",
    "                     header = TRUE, quote = \"\", comment.char = \"\",sep = \"\\t\")\n",
    "head(df.prot, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a1b36781",
   "metadata": {},
   "outputs": [],
   "source": [
    "TMT_columns = seq(15,33,2)\n",
    "dat = df.prot[df.prot$miR.FASP_q.value<0.01,TMT_columns]\n",
    "rownames(dat) = df.prot[df.prot$miR.FASP_q.value<0.01,]$Protein.accession\n",
    "rm(df.prot) # release memory since using dat from here on\n",
    "head(dat, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d870c2fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "dat.log = log2(dat)\n",
    "#remove rows with NAs\n",
    "dat.log = na.omit(dat.log)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ea75031a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a vector with the treatment group in the same order as samples in the protein table.\n",
    "cond = as.factor(c(\"ctrl\",\"miR191\",\"miR372\",\"miR519\",\"ctrl\",\n",
    "\"miR372\",\"miR519\",\"ctrl\",\"miR191\",\"miR372\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cfba8ab",
   "metadata": {},
   "source": [
    "## Protein_intensity table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc1a022f",
   "metadata": {},
   "source": [
    "From /general_analysis/DataBaseUpload/NGS/Proteomics/bin/PhosphoProteomics_PreProccess.html:\n",
    "\n",
    "Outlayiers: very low/high values (0.0001% and 99.9999%) were floored to these ranges. This believed to be an error and unlikely to be real value.\n",
    "\n",
    "Possible reason: the original data have Inf. The R language calculated it to some extreme values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "79372130",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from /general_analysis/DataBaseUpload/NGS/Proteomics/bin/PhosphoProteomics_PreProccess.html\n",
    "# `Counts` are just 2 to the power of Intensity. This is for DE analysis where counts are needed\n",
    "# note: without filtering, the dimension is 1330824 rows\n",
    "intensity <- read.csv(\"../proteomics/data/Protein_intensity.csv\") \n",
    "# %>% filter(Intensity > -3.0) # remove values that are too small - think if this is necessary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "83423d93",
   "metadata": {},
   "outputs": [],
   "source": [
    "int_mtx <- intensity %>% \n",
    "    select(c('ProteinGroup', 'Sample', 'Intensity'))  %>% \n",
    "    pivot_wider(names_from = Sample, values_from = Intensity, values_fn = mean) %>% \n",
    "#     pivot_wider(names_from = Sample, values_from = Intensity) %>% \n",
    "    column_to_rownames('ProteinGroup')\n",
    "# head(int_mtx)\n",
    "# print(dim(int_mtx))\n",
    "# rm(intensity)  # release memory since using int_mtx from here on"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "87879ef6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>295</li>\n",
       "\t<li>317</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 295\n",
       "\\item 317\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 295\n",
       "2. 317\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 295 317"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# If omitting all proteins with NA values, how many rows are left?\n",
    "x <- na.omit(int_mtx)\n",
    "dim(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "30e9560d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.509788795159781"
      ],
      "text/latex": [
       "0.509788795159781"
      ],
      "text/markdown": [
       "0.509788795159781"
      ],
      "text/plain": [
       "[1] 0.5097888"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# proportion of missin values\n",
    "sum(is.na(int_mtx))/(dim(int_mtx)[1] * dim(int_mtx)[2])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18e9e20b",
   "metadata": {},
   "source": [
    "The intensity is already in log-scale because it has negative values."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b14f7cf",
   "metadata": {},
   "source": [
    "### Check the distribution of non-missing values\n",
    "\n",
    "Missing values in proteomic data can be generally characterized into missing at random (MAR) and missing not at random (MNAR). \n",
    "+ MAR missing values mostly result from technical limitations and stochastic fluctuations in an abundance-independent manner.\n",
    "+ MNAR missing values are more abundance-dependent that can be explained by the measurability of the corresponding peptides. \n",
    "\n",
    "Missing values in proteomic data are a mixture of MAR and MNAR. Although the real proportion is difficult to determine, it is believed that MNAR plays a dominant role in producing missing values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "962225f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAMFBMVEUAAABNTU1oaGh8fHyM\njIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD////QFLu4AAAACXBIWXMAABJ0AAAS\ndAHeZh94AAAgAElEQVR4nO3dCXuyvBaF4YBjHeD//9vKHBAUdAE7+NzXOV/tBhKGrCLI27oU\nwNfc2isAbAFBAgQIEiBAkAABggQIECRAgCABAgQJECBIgABBAgQIEiBAkAABggQIECRAgCAB\nAgQJECBIgABBAgQIEiBAkAABggQIECRAgCABAgQJECBIgABBAgQIEiBAkAABggQIECRAgCAB\nAgQJECBIgABBAgQIEiBAkAABggQIECRAgCABAgQJECBIgMCWguSca79qCr7DIitzipxrenqs\nyV/9anwjvYsdH6+PknX0upHON13PMZmvs1kEtbJvjArSNVpkk0+PvttBilqrNk7vYlnL2m1Y\nO0i9x4QgrWZUkBY6PrFzt3av5VlkapCeFvvLg/SnWc26G+l8kv4J0mqegvR6poXWpfrWufvk\n/vsW2z2y5dxOtJ5VN8rmgutfIfwtaAydkZLTY/C5/V9ajsxytsshe/d1KRe5P77bnb0l73F+\nMvjbP17Hx3o0n2MXX9P0HLndtd19q72+IO1bE55mvzzW8uCfxnoXu2fv9uIyXjVvE8esc1l0\nh3trZZNj5KJisU6LT7u1Z2X9Xr0F2v2Ue9Xfeu+Y+CvwujNzfiBI96g8VLvWQduVr/Nxml7L\nWZol43yBai53LarF9/djXav57fl5TZv2rt5Ktrt3+Wkm0xovz4tltxpO2RVY63aDv4lj1jnr\nN/8u8k93VSvXpxafdmvPyrZ6bRbo9FPs1YGd5a/A687s+YEgHfIriuRx7M7eQdtXx70YylH9\nbbWkyxY7P457ko9e75A/xoUfgkKrPa+leoVuj1HUrFqn+6b31v2rp8XyFU3SpL4JUfA3ccw6\nN93F3u6q9kHUbbGzg3tXtt1rs0Cnn3yvDu0sfwVedmbQxoLUKAvFf7Ofh0lrRF4eX8+PEZnd\nXbvkV/BR9iVqlsyGRf0eqm7pUT1n4+KWf2n67rTX89YuGz3nakLP7Fn/hzeL5QtmP9H3xWLe\nfM0mjljnrLvHj/1L5K9sEYWkiFB7p3mrM7Sycd814HM/+V4d2FmtFXjZmUG2126agSBl4agv\nharjc6h+1B7zH3XVwPxrlrx0mi7+e219aWbotNcXpOI04qqzZHf2rL/kzWL1il7ap8P2Jr5f\nZ1fe9rv4K7vPznX5d/uXLQ6tbGue+ptOP5cXO6u1AqM6M8T22k0zEKRT+c7gkja17Et+0LKL\nd5ePm7R38mOGv+PO1YMyffpSL+e31xek4sqmntw7e/1d0WN3seY9XdSs4NMmvl9n12xu3O3V\n5V10W0x7lu6Oba/X9gJeP8mLndVagXedWWN77aZp9nX7IByrw3N/mlS98o54e/Jf7CXzdZBa\nr/qClI3+p5y9D5K3WP7erOJdvLQ38f06dza3GyTXbbF3t3a2sdVrd6/0beVTvb0CBGk1Q0FK\nk7/iLtGuNan+oRj1npHyb7O3evHhfBsTJL+9/iD9FTexXszeG6RmsbQerK518dLaxBHrXHfv\nzxO1V9nfaf27tbWN7V67e+UpSH1b314BgrSawSBlLofWwdy/vUbKp8Zl/W2Q9u+vkdLqru+L\n2d8sdnW+zudY1SaOWGdXdn/xL0c6NzD8nda/W1vT2r02Czz3M7j17RUgSKsZCFJcvzOvfvol\ng3ftXP/he39GGnHXrmwnfzU0+5vFjs2zQWf/o6TWJo5Y56y5v+Ju2rmu/hW32P7yk1B7p/Xt\n1v73tk9npOd++rc+6awAQVrPQJCyO8b3tPoQM7sdlX2tPz8sPp94/hwpb2iXz3yJ3gap215/\nIvK7uO7F7G8Wc80thtZtrNYmjljnZmP9j2zqfXDt7rS+3dpZ2XavzQI9/fTu/OrYVCtAkNYz\nEKT6ujl/t39wnc//i8/5LuUsncNXvZeK8oP7Kkid9gYSkVRBGpj99WJ//qeSe//JVX8TR6yz\nqx4XaD3ZUO6DIjqtnda7W9sr2+61WaCnn6etr45JawUI0mqGglS81d+Vd7n2dXYOkXd/95Y9\na3d5CklWjg63e/UZfE/rlVZ7A4nwPsbtnf31Yjv/IuLSGuT+Jr5f5+zLX+yiY9LqJjk+3tHt\nL88t9u/W9sq2evUW6Omnu/X1MfFXgCCFrPvszUYtNSqtj36hn9nQ11zxPum2az8xsFUESe5n\nNvS15ur36RbwFhEkuZ/Z0NfqfzUg/m0IRhEkuZ/Z0DeSU/bwQPT0nOY2ESS5n9lQYE4ECRAg\nSIAAQQIECBIgQJAAAYIECBAkQIAgAQIECRAgSIAAQQIECBIgQJAAAYIECBAkQIAgAQIECRAg\nSIAAQQIECBIgQJAAAYIECBAkQIAgAQIECRAgSIAAQQIECBIgQJAAAYIECBAkQIAgAQIECRAg\nSIAAQQIECBIgQJAAAYIECBAkQIAgAQIECRAgSIAAQQIECBIgQJAAAYIECBAkQIAgAQIECRAg\nSIAAQQIECBIgQJAAAYIECBAkQIAgAQIECb/A9ZF2oGwMMKpvnBMkYCKCBAgQJECAIAECBAkQ\nIEiAAEECBAgSIECQAAGCBAgQJECAIAECBAkQIEiAAEECBAgSIECQAAGCBAgQJECAIAECBAkQ\nIEiAAEECBAgSIECQAAGCBAgQJECAIAECBAkQIEiAAEECBAgSIECQAAGCBAgQJECAIAECBAkQ\nIEiAgOEgXU97l9kfr8L1AeZgNkhJ7Bo75RoBemaDdHTR3y1/db9E7qhbIWAGZoMUuVv9+uYi\nzcoAMzEbJOeGvgHsMRskzkgIidkgPa6RLvf8FddIsM9skNKdd9cuTpSrBMjZDVJ6PeafI0X7\nE58jwTrDQQLCQZAAAcNB4hEhhMNskHhECCExGyQeEUJIzAaJD2QRErNBevOIkPN92AUgYzZI\nE85IBAmrMxukCY8IESSszmyQJjwiRJCwOrtBGv+IEEHC6gwHyVIXwGsECRCwG6T7wUWnND3H\nLnrzcSxBwurMBimJsguk82nEI0IECaszG6Rjdsv7GLlDkiZHbn/DOLNBivIFnctvfPOBLIwz\nGyTnmv+++S1CBAmrMxukyAtSwhkJxpkNUnWNdEzK1/ouABmzQeKuHUJiNkh8joSQ2A2SqS6A\n1wgSIECQAAGCBAgQJECAIAECBAkQIEiAAEECBAgSIECQAAGCBAgQJECAIAECBAkQIEiAAEEC\nBAgSIECQAAGCBAgQJECAIAECBAkQIEiAAEECBAgSIECQAAGCBAgQJECAIAECBAkQIEiAAEEC\nBAgSIECQAAGCBAgQJECAIAECBAkQIEiAAEECBAgSIECQAAGCBAgQJECAIAECBAkQIEiAAEEC\nBAgSIECQAAGCBAgQJECAIAECBAkQIEiAAEECBAgSIECQAAGCBAgQJECAIAECBAkQIEiAAEEC\nBAgSIECQAAGCBAgQJECAIAECBAkQIEiAAEECBAgSIECQAAGCBAgQJECAIAECBAkQIEiAAEEC\nBAgSIECQAAGCBAgQJECAIAECBAkQIEiAAEECBAgSIECQAAGCBAgQJECAIAECBAkQIEiAAEEC\nBAgSIECQAAHDQbqe9i6zP17n6gIQMRukJHaN3SxdADJmg3R00d8tf3W/RO44RxeAjNkgRe5W\nv765aI4uABmzQXJu6BtZF4CM2SBxRkJIzAbpcY10ueevuEaCfWaDlO68u3ZxMksXgIrdIKXX\nY/45UrQ/8TkSrDMcJEtdAK8RJEDAcJB4RAjhMBskHhFCSMwGiUeEEBKzQeIDWYTEbJDePCLk\nfB92AciYDRJnJITEbJB4RAghMRskHhFCSOwGiUeEEBDDQbLUBfAaQQIE7AYpOWa36k6xc7u/\nmboAVMwG6R45lyYRjwghCGaDdHD75PGfw/2RqQO3v2Gc2SA5l5T/ebzL4wNZGGc4SGn2eIP3\njbwLQMZskA7ZI0Kn4jmh5PVFEkHC6swG6eai4y3dR48kXWJ3maMLQMZskNJL1DwidJqnC0DF\nbpDS9O+Q/yvZ/ek+WxeAhuUgGeoCeI0gAQIECRAgSIAAQQIECBIgQJAAAYIECBAkQIAgAQIE\nCRAgSIAAQQIECBIgQJAAAYIECBAkQIAgAQIECRAgSIAAQQIECBIgQJAAgZ8Lkusz15rhZ/xe\nkL5tAOhBkKY2APQgSFMbAHoQpKkNAD0I0tQGgB4EaWoDQA+CNLUBoAdBmtoA0IMgTW0A6EGQ\npjYA9CBIUxsAehCkqQ0APQjS1AaAHgRpagNAj0WDFJ/uyqb7uvhoZoKEby0aJOfcHFkiSFjd\nokFK/g5zZIkgYXWLXyNdT7E6SwQJq1vjZsMtepyXzrN2MW1mgoRvrRCkyy7/PQm7GbuYODNB\nwreWDlJyepyO4kvySNN+pi6mz0yQ8K1lg3TNbjYcb8UEWTcECatb9nOkx8nonFQTojm6+Ghm\ngoRvLfs50v6ibLqvi/aEkb/DjiDhW8t+jqRsuL+LtxMIEuaw7DVScszez0VHbaIIEha1yjsd\nv7F7lPfnXCR9toEgYVGrjCu/sZ07ZOei5Ki79d3t4u0EgoRvrR6k+vyn/bX1BAmLWj1IkSsu\njhKChICtHqSj210fX647d5yri7cTCBK+tXqQ0l15h0P3nN1TF+8mECR8a/0gpX/7LEbCJ7+f\nu3gzgSDhWwaCNAuChEURpAk1YAhBmlADhqwfpOyfmev//DFBwqJWD9Jpnr8jTpCwqNWDFCl/\nU0N/F28nECR8a/UgaU9EvV28nUCQ8K3Vg7R3s/yLJIKERa0epHuUPyKkRpCwqNWD9PqfQUm6\neDuBIOFbBGlCDRiyepBmQpCwKII0oQYMMRCkyz57V7fX/jkKgoRFrR+kXXF5xC8/QchWD9LZ\n7fJ/ZX52h7m6eDuBIOFbqwcp+50N5S/kmquLtxMIEr61epDyt3UECYFbPUhxeUa6uXiuLt5O\nIEj41upBKq+RLuKnwAkSFrV6kNI9v0UI4Vs/SPnnSG7/p+yAIGFhBoI0C4KERRGkCTVgCEGa\nUAOGrB4k/hkFtoAgTagBQ1YPUum6k/6dMYKEZVkJUprw0CoCZiZIPGuHkJkJ0tlFc3cxOIEg\n4VurB6m513Caq4u3E3prI//cO5AxE6RY+5uLZzkjESQMWT1IMyFIWBRB+rIGZFYP0kxXIQQJ\niyJIX9aAzOpBSk/R5fHfaxTAP+wjSBiyepBO7pZ/vTnpM0IECYtaPUj1u7kAnmwgSBiyepCi\n+oxk/7cIESQMWT1IR5dfI438LULXU/G7UvbHN3+djCBhUasHqfjd3w/H98slsXd/7/XNCYKE\nRa0fpPQv/y1ClxHLPc5ef8UbwfvjDPYyeQQJizIQpPGq66nM7fXT4gQJiwoqSK0be6/v8hEk\nLMpAkMb/obHVz0j80woMWD9IE/7QWHaHr5jLzjUSQUJm9SBN+kNjO+9MECdju3g7gSDhW6sH\nadofGrse88+Rov3JyudIBAmZ1YMU0h8aI0gYsnqQQvpDYwQJQ1YP0rQ/NGbvESGChMzqQZry\nh8YsPiJEkJBZP0gT/tCYxUeECBIyBoI03uofyPbV+JAWqYEg7Uc89V0v9/oRoVFjeYkzEkH6\nPasHacqPb5NnpCmdY7NWD1J2+3usUB4RIki/Z/UgJfvdmzvZnkAeESJIv2f1IE27RA/jESGC\n9HsCC9JHXbydQJDwrdWDNBOChEUFGqS3py+ChEWtGqTP380RJNhiIEjj4zTh+QGChEUFFaRr\nRJBgU1BBSpO9291HLUOQsKiwgpSmf879jVmGIGFRoQUpve/cPiFIMCa4IGV/USm6ECTYsnKQ\nPvpXPLf4/cwECYsKMUhpeiBIsGXVIM2IIGFRBGmhGraNIC1Uw7YRpIVq2DaCtFAN20aQFqph\n2wjSQjVsG0FaqIZtI0gL1bBtBGmhGraNIC1Uw7YRpIVq2DaCtFAN20aQFqph2wjSQjVsG0Fa\nqIZtI0gL1bBtBGmhGraNIC1Uw7YRpIVq2DaCtFAN20aQFqph2wjSQjVsG0FaqIbtcH365htZ\n+3xFlI1N7IIg4VtmxgZBQsjMjA2ChJCZGRsECSEzMzYIEkJmZmwQJITMzNggSAiZmbFBkBAy\nM2ODICFkZsYGQULIzIwNgoSQmRkbBAkhMzM2CBJCZmZsECSEzMzYIEgImZmxQZAQMjNjgyAh\nZGbGBkFCyMyMDYKEkJkZGwQJITMzNggSQmZmbBAkhMzM2CBICJmZsUGQEDIzY4MgIWRmxsbm\ngzTyN3EiTARpYAJnKUxhZhwQJITMzDggSAiZmXFAkBAyM+OAICFkZsYBQULIzIwDgoSQmRkH\nBAkhMzMOCBJCZmYcECSEzMw4IEgImZlxQJAQMjPjgCAhZGbGAUFCyMyMA4KEkJkZBwQJITMz\nDggSQmZmHBAkhMzMOCBICJmZcUCQEDIz44AgIWRmxgFBQsjMjAOChJCZGQcECSEzMw4IEkJm\nZhwQJITMzDggSAiZmXFAkBAyM+OAICFkZsYBQULIzIwDgoSQmRkHBAkhMzMOCBJCZmYc/GSQ\n+Ct+m0GQBiZYqsE+M8ecIA3XYJ+ZY06Qhmuwz8wxJ0jDNdhn5pgTpOEa7DNzzAnScA32mTnm\nBGm4BvvMHHOCNFyDfWaOOUEarsE+M8ecIA3XYMvYJ1IIkrEabDF9zAnScA22mD7mBGm4BltM\nH3OCNFyDLaaP+eeNXU/7/Gpvf7x+2IWl0BAk+0wf808bS2Lvzsnusy4shYYg2Wf6mH/a2NFF\nf7f81f0SueNHXVgKzehbq/yjwNVsMkiRu9Wvby76qAtTQRLXoGf6uH3aWOvH8OufyQQJEqaP\nG2ekeWrf4O1jP9PH7YtrpMs9f7WRayRx7Ruc9fqZPm4fN7bzflrGyUddWBr4Zg7IQn2EyPRx\n+7yx6zH/HCnan7b6ORJBssX0cVvgABGkiQhSP9PHjSAJauKbAwSp30aDtP1HhNY6SASpn6Vj\nJGvsFx4RIki2WDpGssY294hQX23sY0Mja2MRpH6bDBIfyE6vjUWQ+lk6RrLG3jwiNOq629Ig\nt3SQCFI/S8dI1hhnpOm1sQhSP0vHSNYYjwhNr41FkPpZOka6xnhEaHJtLILUz9IxEjbGI0JT\na2NtJUhrfVAdWJC+78LSILd0kDYTpJE1dXsEaeO1sX4tSL1nLvFneHaDlByc213KRl62QpAm\n+rkgGap97NPGkqh40K5ohCCNqY1FkOwfI1ljR3d+pOkc5Y/ZEaRRtbEIkv1jJGssKha8R/Gd\nII2sjUWQ7B8jWWNVdpLdjiCNrPUZfVEtvpW8BEv73myQYld9CBvvCNK4Wh8zA2EG3/xAsHSM\nRvm0sbM7lK/ubkeQRtX6mBkIM7C0780GKT3W6bm8eZNBkF4wMxBmYGnf2w1SettXr+4HgvRU\nC+0DxRmYOh4jax9b4GD8ZpD6aguEyxJT+35k7WMEaVs1SyztF4JEbVLNEkv7hSBRm1SzxNJ+\nIUjUJtUssbRfCBK1aTVDT0CY2i8jax8jSL9ZW4Kl7SVI1GapLcHS9hIkarPUxvrmraKl7SVI\n1GapjfVNe5a2lyBRm6U2FkEaiSD9Zm0sgjQSQfrN2lgEaSSC9Ju1sQjSSATpN2tjEaSRCNJv\n1sbqbc/4PxcnSNQWq41laZ0t7Zd5G5vYhaUd+HM19VllI2epjxEkatQECFLAtY2cBSzVPkaQ\nqFETIEjUqAkQJGrUBAjSD9S4liJI1KgtWfsYQaJGTYAgUaMmQJCoURMgSNSoCRAkatQECBK1\nusZt8s8RJGrUBAgSNWoCBIkaNQGCRI2aAEGiRk2AIFGjJkCQqFETIEjUqAkQJGrUBAgSNWoC\nBIkaNQGCRI2aAEGiRk2AIFGjJkCQqFETIEjUqAkQJGrUBAgSNWoCBIkaNQGCRI2aAEGiRk2A\nIFGjJkCQqFETIEjUqAkQJGrUBAgSNWoCBIkaNQGCRI2aAEGiRk2AIFGjJkCQqFETIEjUqAkQ\nJGrUBAgSNWoCBIkaNQGCRI2aAEGiRk2AIFGjJkCQqFETIEjUqAkQJGrUBAgSNWoCBIkaNQGC\nRI2aAEGiRk2AIFGjJkCQqFETIEjUqAkQJGrUBAgSNWoCBIkaNQGCRI2aAEGiRk2AIFGjJkCQ\nqFETIEjUqAkQJGrUBAgSNWoCBIkaNQGCRI2aAEGiRk2AIFGjJkCQqFETIEjUqAkQJGrUBAgS\nNWoCBIkaNQGCRI2aAEGiRk2AIFGjJkCQqFETIEjUqAkQJGrUBAgSNWoCBIkaNQGCRI2aAEGi\nRk2AIFGjJvB5Y9fT3mX2x+uHXVjagdSofeXTxpLYNXafdWFpB1Kj9pVPGzu66O+Wv7pfInf8\nqAtLO5Aata982ljkbvXrm4s+6sLSDqRG7SufNubc0DdlxTPcBrCiD8d+/2D+cLkJZyRg+764\nRrrc81dvr5GA7fv49LbzTpFxolwlIDyfv0+8HvPPkaL96c3nSMD2SS+4gF9FkAABggQIECRA\ngCABAgQJECBIgABBAgQIEiBAkAABggQIECRAgCABAgQJECBIgABBAgQIEiBAkACBNYO00i9h\nwtatMpjX6HTBvumDPhZBkOiDPkLtdMG+6YM+FkGQ6IM+Qu10wb7pgz4WQZDogz5C7XTBvumD\nPhZBkOiDPkLtdMG+6YM+FkGQ6IM+Qu10wb7pgz4WQZDogz5C7RTYGoIECBAkQIAgAQIECRAg\nSIAAQQIECBIgQJAAAYIECBAkQIAgAQIECRAgSIAAQQIECBIgsFqQjpGLjslcrfu/TX2Wns7V\njvNaV3dU9THfxpzjvpWfq4/ZtiM5OHe4pd2G5x1iHWsFaZfv03im1m/eMZulp1v1Jw+81tUd\nVX3MtzHHvLUoSWfcjqaP+bYjylu7dRqed4h1rRSkq4tu6S1y13mav7n9rD092nPd1tUd1X3M\ntjE3d0iy895hxu3w+phtO45Z68e89RmPxxsrBenoLo///rnTPM2fm4bn6OnsdtU7lKZ1cUdN\nH7NtzL5oP+tmtu3w+phtOyKXlF3MeDzeWSlIe3dPWz+jtM7uPGdP7piWg9xrXdxR08fMG5N3\nM992NH3MfVCidP7teNH9Mt08dev8L3J7dzk8LjTn6unWbTb7Iu6o6WPmjUncbs7taPqYdzuO\neUxn3o4Xthqk3G6+nmYPUuoFadaNOWdvgWYegHkfc27Hn3PdhBIkSfN/jx+D+Q+p8IM078bc\no3069wCs+phvO877KL8YIkhzSLJ7n+EHqTDTxiTRzmtxnu0o+yi/meugHDoJ/Y0gRYtsZdb8\nTD2V7Xmt6ztqNzVPH7vic5ZZt2PX+ixnroOSZHcbZt2Ol1a9a3ef+ZZKcz9K3lPrrt29uUuk\n7Og5SOo+7vHunr+YcTvqPkqzHZROw8sMsab3ZbrpOuU3+S/FFaJe8clCvhNn6qkc5F7r+o7q\ns95cG3Nx1Xuu+baj6WO27agajmc+Hi+tFKSZP3Y+ZrsvyT+Sm6mn+Z9sqPuYbWPu9Rifbzu8\nPmbbjvzJhmSfXSP93JMNaVzfCp1DUjx9dZyvp+ptl9e6vKOyj9k25uCap9/m2g6vj/kOStS7\n8vMOsa61gpTkj+bO2nx8nrGnKkhe6/KO/D7m2BjXDPLZtqPbxzwHpbfhmYdYx1pBAjaFIAEC\nBAkQIEiAAEECBAgSIECQAAGCBAgQJECAIAECBAkQIEiAAEECBAgSIECQAAGCBAgQJECAIAEC\nBAkQIEiAAEECBAgSIECQAAGCBAgQJECAIAECBAkQIEiAAEECBAgSIECQAAGCBAgQJEDgh4Lk\n8r/Om777i/ET/558cnCf/r3fdz1dhme6jG9FsCLKvi6d7wVNmrCRzRjDuah68Xq2Sa3unXOn\nT1fo5eTYDc4Ue8WwghR3WyBIwalHvDRIzt0/X6ERk3tncisF6Xtbyc2TrW5XD+fiYtCLg/TF\nCo2YTJDCsNXt6uHcze2LF/4offz/5KLHuepYXOw8vj/Wfw37HLvoXMyZxMXSVT3/M9r1X+zO\nXXbO7fKLgMvjHV/RxovmiwXrLrzF06rpav5Tq9FWpz0zJC7OJ8Uu8dtP21Na69iK7dAqlTPe\n92V/5VbX7fgrdGzttdb+6uzYniYD9FNBSg/umj4H6ZQd3WzI5EPdueyyx+2y6fmr/GVerUfL\nrqq3gnQuvnuMj1Pxqhj03eZPTZutLprFy9Utg7R/arQTpOcZdvmp954127Rfrnk1pbWOrSAN\nrlI5Y+T860KvnWombwv35fr4+6uzY5+bDNFvBan4gdwN0i7JRkz+3yg/qrf0Frm/x8/arJjs\nsrt9+fTKXzOL/14lcrdsWpwV/7JXLn3ZfDbd66JZvF7fZvluo/5mPc/wlw/K06NVr/1y1asp\nneaaHfJmlZr+qlrdTlXwtjDfa3/tbe7s2OcmQ/RbQXocrfNzkIqz1L3+Pht0l+wdyT57a/RI\n376aq7QvZ6lPK1UH7Xu75SAdbD77ttVFz+L18p1G/dn6ZsgGZdxuv5xcTek01+yQN6vU7c9f\nWW83XLy95u2vYr5WF0NNhiXolZ8mP1DZxcHTNVL6/H350lXv3VpH2ZvFn/C4CtrfbsXr++W0\n86YPNN/qwl887VnyqdHn2eoZDo/g3ut3gf47wXpKpzl/DV+u0vM9kLqdp73j76j20q0d+9xk\neIJe+WnyA3V1h9mClJ6yt/pRdu7Z1aN3fJC8xdO0Z8mnRp9ma2a4Pt7BHbOf809Bqqd0musN\nUt8qPY36XbsDgrR11buK24QgdRZuf/M0ph9vX45x9t7p4OLz5T4uSD2L+700/31qtDubN0Ma\nxdn/+kZnNaXT3NMaDq1Sd9T7vXb3znCQWnMTpKCU735cXB+7a3+Qsh/X5TXSpb1wad9cBTwf\n/rrNoSAVzR+Kb/edqxCvvd4IvgqSN8PjpHPObys8tV9P6TTX7JA3q9Qd9X6vZaG1he39Vf00\nu7xqMkRBr/w05YE65e8oYnfObhv1Bqm4xXQp7zal52YAlAbu2sXF/au4GEu3oWukuvnsW69p\nFBgAAAGNSURBVK+LZvFqfe+dJb1Gm8cpemfIRrarb5gV7ZeqKZ11bHbIy1XqC1LTa1lobWHa\nuWt3T592LEEKSnWgouxF/gnJvj9Ih3xa9n3x9j+7Rmgf5fpzkdaEv+J9/zW/SK9ePje/q5rP\ni00XzeKF2BW3y6slvUbj+rnBgRmyWXbeqkbec0zllM46NjvkxSr1jfpWr8Wk9ha29le55u0d\nS5CCUh2oS/7icRndve1QH9Vj/TH7+XHgD/f06Sifo+KT+vaE/DmAfEAdshfee5lW8/ty2aJY\nd+EtnrvG7SB5jRaT/M3qzJBl4NLZhEo1pbOO9Q55sUq9o97vtZzU3kJ/f1Vr3tqxBAmThT1c\nxtj+Fvb5yY1e0/aH2fa3sM9PbvSatj/Mtr+FfX5yo9e0/WG2/S3s85MbDagRJECAIAECBAkQ\nIEiAAEECBAgSIECQAAGCBAgQJECAIAECBAkQIEiAAEECBAgSIECQAAGCBAgQJECAIAECBAkQ\nIEiAAEECBAgSIECQAAGCBAgQJECAIAEC/3dVzR2Gy8W9AAAAAElFTkSuQmCC",
      "text/plain": [
       "Plot with title \"Histogram of Non-NA samples in a protein\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# For each protein group, how many samples have intensity values\n",
    "hist(rowSums(!is.na(int_mtx)), breaks = 30,\n",
    "     main = \"Histogram of Non-NA samples in a protein\",\n",
    "     xlab = \"Number of samples that have values in a protein\")  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "71742a2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each sample, how many protein groups have intensity values\n",
    "hist(colSums(!is.na(int_mtx)), breaks = 18,\n",
    "     main = \"Histogram of Non-NA proteins in a sample\",\n",
    "     xlab = \"Number of proteins that have values in a sample\")  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f656dda",
   "metadata": {},
   "source": [
    "### Filtering, normalization, and imputation\n",
    "\n",
    "First, filter proteins with too many missing values. \n",
    "\n",
    "Seoncd, we need to account for technical variability in the amount of sample analyzed by the mass spectrometer from one run to another. This is an issue parallel to the variation in sequencing depth in RNAseq experiments. To normalize out these technical differences, we perform a median normalization. For each sample, the median of the log2-transformed distribution is subtracted from all the values.\n",
    "\n",
    "Third, since missing values are associated with proteins with low levels of expression, we can substitute the missing values with numbers that are considered “small” in each sample. We can define this statistically by drawing from a normal distribution with a mean that is down-shifted from the sample mean and a standard deviation that is a fraction of the standard deviation of the sample distribution. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9d55571c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This plot is to check if the data are currently median centered\n",
    "boxplot(int_mtx, las=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f058861",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep the proteins that were at least observed N_miss times\n",
    "N_miss <- 80\n",
    "int_norm <- int_mtx[rowSums(!is.na(int_mtx)) > N_miss,]\n",
    "print(nrow(int_norm))\n",
    "# proportion of missing values\n",
    "print(sum(is.na(int_norm))/(dim(int_norm)[1] * dim(int_norm)[2]))\n",
    "# # Bring the proprotion of missing values to below 30%\n",
    "\n",
    "# Normalization to conteract the different median between samples\n",
    "# subtract each value in a col with its respective col median\n",
    "int_norm <- sweep(int_norm, 2, apply(int_norm, 2, function(x){median(x, na.rm = TRUE)}), '-') \n",
    "\n",
    "# hist(unlist(int_norm), breaks = 30, col = \"blue\",\n",
    "#                   main = \"Histogram of protein intensity\", \n",
    "#                   xlab = \"Intensity\")  \n",
    "\n",
    "# Imputation\n",
    "width <- 0.3\n",
    "downshift <- 1.8 # assume missing values (small values) are outside 90% of the values in a normal distribution\n",
    "# shrink sd width\n",
    "temp_sd <- width * apply(int_norm, 2, function(x){sd(x, na.rm = TRUE)}) \n",
    "# shift mean of imputed values\n",
    "temp_mean <- apply(int_norm, 2, \n",
    "                   function(x){mean(x, na.rm = TRUE)}) - downshift * apply(int_norm, 2, \n",
    "                                                                           function(x){sd(x, na.rm = TRUE)})  \n",
    "set.seed(1)  # set seed for rnorm()\n",
    "for (i in 1:(dim(int_norm)[2])) {\n",
    "    temp <- int_norm[, i]\n",
    "    int_norm[, i][is.na(temp)] <- rnorm(sum(is.na(temp)), \n",
    "                                             mean = temp_mean[i], \n",
    "                                             sd = temp_sd[i])\n",
    "}\n",
    "head(int_norm, 2)\n",
    "print(dim(int_norm))\n",
    "# hist(unlist(int_norm), breaks = 30, col = alpha(\"green\", 0.5), add = TRUE)\n",
    "# legend('right', c('Normalized', 'Imputed'), fill = c(\"blue\", alpha(\"green\", 0.5)))\n",
    "\n",
    "# The imputed has both the normalized and the new imputed values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8d51152",
   "metadata": {},
   "source": [
    "### Differential protein expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd5f1d5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "clin_treatFile = './data/meta.csv'\n",
    "clin_treat <- fread(clin_treatFile)\n",
    "clin_treat <- clin_treat[clin_treat$Model %in% colnames(int_norm)] %>% \n",
    "    unique\n",
    "head(clin_treat, 2)\n",
    "print(dim(clin_treat))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "59a3e58d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "164"
      ],
      "text/latex": [
       "164"
      ],
      "text/markdown": [
       "164"
      ],
      "text/plain": [
       "[1] 164"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "n_distinct(clin_treat$Model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "99bdb2a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>Tumor type</th><th scope=col>treatment</th><th scope=col>N</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>HEAD AND NECK             </td><td>5-FLUOROURACIL            </td><td>1                         </td></tr>\n",
       "\t<tr><td>HEAD AND NECK             </td><td>5-FLUOROURACIL/CARBOPLATIN</td><td>1                         </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lll}\n",
       " Tumor type & treatment & N\\\\\n",
       "\\hline\n",
       "\t HEAD AND NECK              & 5-FLUOROURACIL             & 1                         \\\\\n",
       "\t HEAD AND NECK              & 5-FLUOROURACIL/CARBOPLATIN & 1                         \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| Tumor type | treatment | N |\n",
       "|---|---|---|\n",
       "| HEAD AND NECK              | 5-FLUOROURACIL             | 1                          |\n",
       "| HEAD AND NECK              | 5-FLUOROURACIL/CARBOPLATIN | 1                          |\n",
       "\n"
      ],
      "text/plain": [
       "  Tumor type    treatment                  N\n",
       "1 HEAD AND NECK 5-FLUOROURACIL             1\n",
       "2 HEAD AND NECK 5-FLUOROURACIL/CARBOPLATIN 1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ct_sum <- clin_treat %>% \n",
    "    group_by(`Tumor type`, treatment) %>%\n",
    "    summarise(N = n_distinct(Model), .groups = 'drop')\n",
    "head(ct_sum, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6fb68e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ct_sum[ct_sum$treatment=='NONE',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6c298c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique(clin_treat$`Tumor type`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40ff9107",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for int_norm matrix, only keep samples that are in clin_treat (name it dat.log to use example code)\n",
    "# add a column in clin_treat to convert treatment to NONE=Naive and all drugs=Treated\n",
    "# sort Model in clin_treat in the same order as in the proteomics matrix\n",
    "# write a function that allows choosing `Tumor type` for each analysis\n",
    "dat.log <- int_norm %>% \n",
    "    select(clin_treat$Model)\n",
    "head(dat.log)\n",
    "print(dim(dat.log))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f94d1bfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "treated_naive <- clin_treat %>% \n",
    "    mutate(treatment = replace(treatment, treatment != 'NONE', 'Treated')) %>% \n",
    "    mutate(treatment = replace(treatment, treatment == 'NONE', 'Naive')) %>% \n",
    "    select(-outcome) %>% \n",
    "    unique\n",
    "print(dim(treated_naive))\n",
    "head(treated_naive)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9270acf5",
   "metadata": {},
   "source": [
    "## Phosphorylation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "23149ea5",
   "metadata": {},
   "outputs": [],
   "source": [
    "phos <- read.csv(\"../proteomics/data/Protein_phos_data.csv\")\n",
    "head(phos, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "50716f1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "phos_int <- phos %>% \n",
    "    select(c('ProteinGroup', 'Sample', 'Counts'))  %>% \n",
    "    pivot_wider(names_from = Sample, values_from = Counts, values_fn = mean) %>% \n",
    "#     pivot_wider(names_from = Sample, values_from = Intensity) %>% \n",
    "    column_to_rownames('ProteinGroup')\n",
    "rm(phos)\n",
    "head(phos_int, 2)\n",
    "print(dim(phos_int))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7f449d9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# x <- na.omit(phos_int)\n",
    "# dim(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2237ac85",
   "metadata": {},
   "outputs": [],
   "source": [
    "# proportion of missin values\n",
    "sum(is.na(phos_int))/(dim(phos_int)[1] * dim(phos_int)[2])"
   ]
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "ir",
   "name": "r-cpu.4-2.m102",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/r-cpu.4-2:m102"
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
